calvin/runtime/south/plugins/transports/calvinfcm/twisted/twisted_transport.py

```diff
@@ -0,0 +1,256 @@
+        if secret == None:
+            _log.error("No secret specified")
+            return
+        authenticator = client.XMPPAuthenticator(jid, secret)
+
+        callbacks = {'connected': [CalvinCB(self._connected)]}
+        xmppfactory = CalvinXMLStreamFactory(authenticator, callbacks, self, self._uri)
+        xmppfactory.addBootstrap(xmlstream.STREAM_AUTHD_EVENT, self.authenticated)
+        
+        self.protocol_factory = xmppfactory
+
+        cf = ssl.ClientContextFactory()
+        reactor.connectSSL("gcm-preprod.googleapis.com", 5236, xmppfactory, cf)
+    
+    def stop(self):
+        traceback.print_stack()
+
+    def authenticated(self, xs):
+        _log.info("Authenticated with FCM server")
+        self._callback_execute('server_started', self._port)
+
+    def stop(self):
+        _log.debug("Stopping fcm server")
+
+    def is_listening(self):
+        return True
+
+    def _connected(self, proto, token):
+        self._callback_execute('client_connected', create_uri(self._jid, token), proto)
+
+# Client
+class TwistedCalvinTransportClient(base_transport.CalvinTransportBase):
+    def __init__(self, host, port, callbacks=None, proto=None, node_name=None, server_node_name=None, *args, **kwargs):
+        super(TwistedCalvinTransportClient, self).__init__(host, port, callbacks=callbacks)
+        self._jid = host
+        self._token = port
+        self._proto = proto
+        self._uri = proto._uri
+        self._factory = None
+        self._node_name = node_name
+        self._server_node_name=server_node_name
+        self._runtime_credentials = None
+        self._is_connected = False
+        if proto:
+            proto.callback_register('connected', CalvinCB(self._connected))
+            proto.callback_register('disconnected', CalvinCB(self._disconnected))
+            proto.callback_register('data', CalvinCB(self._data))
+
+        self._callbacks = callbacks
+        self.sendConnected()
+
+    def is_connected(self):
+        return self._is_connected
+    
+    def sendConnected(self):
+        d = {
+                "connected": 1
+                }
+        self._proto.sendCalvinMsg(self._token, json.dumps(d), msgType="set_connect")
+
+    def join(self):
+        raise NotImplementedError
+    
+    def _connected(self, proto, token):
+        self._is_connected = True
+        self._callback_execute('connected')
+
+    def _disconnected(self, reason):
+        self._callback_execute('disconnected', reason)
+        self._is_connected = False
+
+    def _data(self, data):
+        try:
+            sender = data["from"]
+            if sender == self._token: # MUX
+                payload = data["data"]["payload"]
+                # Payload will be base64 RFC3548 encoded
+                payload = base64.b64decode(payload)
+                self._callback_execute('data', payload)
+            else:
+                _log.debug("Message not meant for me, ignoring")
```

Comment:
yes, this should happen if there are multiple mobile devices connected over fcm to the same base runtime.